cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -517,36 +524,52 @@ public void upgradeApplications(FullHttpRequest request, HttpResponder responder
                                   @PathParam("namespace-id") String namespaceId,
                                   @QueryParam("artifactScope") Set<String> artifactScopes,
                                   @QueryParam("allowSnapshot") boolean allowSnapshot) throws Exception {
-    // TODO: (CDAP-16910) Improve batch API performance as each application upgrade is an event independent of each
-    //  other.
-
     List<ApplicationId> appIds = decodeAndValidateBatchApplicationRecord(validateNamespace(namespaceId), request);
     Set<ArtifactScope> allowedArtifactScopes = getArtifactScopes(artifactScopes);
     try (ChunkResponder chunkResponder = responder.sendChunkStart(HttpResponseStatus.OK)) {
       ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
       try (JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8))) {
         jsonWriter.beginArray();
+        List<Future<?>> futureList = new ArrayList<>();
+        Lock lock = new ReentrantLock();
+        ExecutorService executor = Executors.newFixedThreadPool(UPGRADE_THREAD_NUMBER,
+                                                                Executors.defaultThreadFactory());
```

Comment:
Default thread factory doesn't provide a recognizable name to the threads it creates. Use `Threads.createDaemonThreadFactory` instead.